metricbeat/module/vsphere/virtualmachine/virtualmachine.go

```diff
@@ -104,110 +123,174 @@ func (m *MetricSet) Fetch(ctx context.Context, reporter mb.ReporterV2) error {
+	}
 
-		totalCPU := vm.Summary.Config.CpuReservation
-		if totalCPU > 0 {
-			freeCPU := totalCPU - usedCPU
-			// Avoid negative values if reported used CPU is slightly over total configured.
-			if freeCPU < 0 {
-				freeCPU = 0
-			}
-			event.Put("cpu.total.mhz", totalCPU)
-			event.Put("cpu.free.mhz", freeCPU)
-		}
+	querySpec := types.PerfQuerySpec{
+		IntervalId: 20, // this likely needs to be made into a variable to support other-than-realtime metrics but it's not clear how to define that in the config and map to this and Level correctly
+		MaxSample:  1,
+	}
 
-		totalMemory := int64(vm.Summary.Config.MemorySizeMB) * 1024 * 1024
-		if totalMemory > 0 {
-			freeMemory := totalMemory - usedMemory
-			// Avoid negative values if reported used memory is slightly over total configured.
-			if freeMemory < 0 {
-				freeMemory = 0
-			}
-			event.Put("memory.total.guest.bytes", totalMemory)
-			event.Put("memory.free.guest.bytes", freeMemory)
-		}
+	// Create a slice to hold our query specifications for each VM
+	var metricsToQuery []string
+
+	// Iterate over the counterInfo slice
+	for _, counter := range counterInfo {
+		// Check if the counter is collected at a 20-second interval
+		if counter.Level == int32(m.MetricLevel) {
+			// Add the counter to the metricsToQuery slice
+			metricsToQuery = append(metricsToQuery, counter.Name())
 
-		if host := vm.Summary.Runtime.Host; host != nil {
-			event["host.id"] = host.Value
-			hostSystem, err := getHostSystem(ctx, c, host.Reference())
-			if err == nil {
-				event["host.hostname"] = hostSystem.Summary.Config.Name
-			} else {
-				m.Logger().Debug(err.Error())
-			}
-		} else {
-			m.Logger().Debug("'Host', 'Runtime' or 'Summary' data not found. This is either a parsing error " +
-				"from vsphere library, an error trying to reach host/guest or incomplete information returned " +
-				"from host/guest")
 		}
+	}
 
-		// Get custom fields (attributes) values if get_custom_fields is true.
-		if m.GetCustomFields && vm.Summary.CustomValue != nil {
-			customFields := getCustomFields(vm.Summary.CustomValue, customFieldsMap)
+	vmList, err := m.getVirtualMachineList(ctx, c, mgr, rpList)
+	if err != nil {
+		return fmt.Errorf("error getting Virtual Machine List: %w", err)
+	}
 
-			if len(customFields) > 0 {
-				event["custom_fields"] = customFields
-			}
-		} else {
-			m.Logger().Debug("custom fields not activated or custom values not found/parse in Summary data. This " +
-				"is either a parsing error from vsphere library, an error trying to reach host/guest or incomplete " +
-				"information returned from host/guest")
+	vmChunks := m.getVmChunks(vmList)
+	vmMap := getVmMap(vmList)
+
+	for _, chunk := range vmChunks {
+		// Perform the query for all VMs at once
+		metricBase, err := perfManager.SampleByName(ctx, querySpec, metricsToQuery, chunk)
+		if err != nil {
+			fmt.Printf("Error retrieving metrics: %s\n", err)
+			return nil // this probably needs to be a different error
```

Comment:
why not return error here?

Reply:
Had this in place early in my exploration of editing the module and was not sure if returning error would have some adverse impact, updated.